src/basic/terminal-util.c

```diff
@@ -1206,38 +1208,67 @@ bool terminal_is_dumb(void) {
-         * we are PID 1 then we do not check whether we are connected to a TTY, because we don't keep /dev/console open
-         * continuously due to fear of SAK, and hence things are a bit weird. */
-
-        if (cached_colors_enabled < 0) {
-                int val;
+static ColorMode parse_systemd_colors(void) {
+        const char *e;
+        int r;
 
-                val = getenv_bool("SYSTEMD_COLORS");
-                if (val >= 0)
-                        cached_colors_enabled = val;
+        e = getenv("SYSTEMD_COLORS");
+        if (!e)
+                return _COLOR_INVALID;
+        if (streq(e, "16"))
+                return COLOR_16;
+        if (streq(e, "256"))
+                return COLOR_256;
+        r = parse_boolean(e);
+        if (r >= 0)
+                return r > 0 ? COLOR_ON : COLOR_OFF;
+        return _COLOR_INVALID;
+}
 
+ColorMode get_color_mode(void) {
+
+        /* Returns the mode used to choose output colors. The possible modes are COLOR_OFF for no colors,
+         * COLOR_16 for only the base 16 ANSI colors, COLOR_256 for more colors and COLOR_ON for unrestricted
+         * color output. For that we check $SYSTEMD_COLORS first (which is the explicit way to
+         * change the mode). If that didn't work we turn colors off unless we are on a TTY. And if we are on a TTY
+         * we turn it off if $TERM is set to "dumb". There's one special tweak though: if we are PID 1 then we do not
+         * check whether we are connected to a TTY, because we don't keep /dev/console open continuously due to fear
+         * of SAK, and hence things are a bit weird. */
+        ColorMode m;
+
+        if (cached_color_mode < 0) {
+                m = parse_systemd_colors();
+                if (m >= 0)
+                        cached_color_mode = m;
                 else if (getenv("NO_COLOR"))
                         /* We only check for the presence of the variable; value is ignored. */
-                        cached_colors_enabled = false;
+                        cached_color_mode = COLOR_OFF;
 
                 else if (getpid_cached() == 1)
-                        /* PID1 outputs to the console without holding it open all the time */
-                        cached_colors_enabled = !getenv_terminal_is_dumb();
+                        /* PID1 outputs to the console without holding it open all the time.
+                         * Also note the Linux console can only handle 16 colors.
+                         */
+                        cached_color_mode = getenv_terminal_is_dumb() ? COLOR_OFF : COLOR_16;
                 else
-                        cached_colors_enabled = !terminal_is_dumb();
+                        cached_color_mode = terminal_is_dumb() ? COLOR_OFF : COLOR_256;
         }
 
+        return cached_color_mode;
+}
+
+bool colors_enabled(void) {
+
+        /* Returns true if colors are considered supported on our stdout. */
+
+        if (cached_colors_enabled < 0)
```

Comment:
actually, I think we shoul drop this cache. I mean, it's gonna be in sync with cached_color_mode always anyway, thus we might just make `colors_enabled()` an alias for `get_color_mode() != COLOR_OFF` via inline func or even macro and runtime behaviour should be pretty much the same

Reply:
I defined an inline version of the function in `src/basic/terminal-util.h` and it seems to work.